### PR TITLE
Api Library 2.0 Support

### LINKED ISSUE #134: Update to 2.0 version of api library
2.0 version of the api library is now released.

It is mostly not backwards compatible.

Woo plugin should use new $config format in the constructor so that whitelabel version of the plugin can be easily configured to use different config.

Library also has getToken() and setAccessToken() functions in Client class. When token authentication is enabled, token should be cached for the time it is alive.

### PR DESCRIPTION
**Pull request by joosev**
Resolves #134

### CODE REVIEW

**Comment by joosev** [comment]
I don't understand why it fails.

**Comment by joosev** [comment]
I was able to fix the original error, but because we don't have any conatiner/docker image for running tests - it is too difficult to setup testing environment. See: #146 

I don't know why search pickup points unit test fails.

**Comment by k1sul1** [comment]
It fails because shipment->get_pickup_points() throws an Exception, which is not catched. 
```

// at end of wc-pakettikauppa.php
add_action('wp_loaded', function() use ($instance) {
  error_log($instance->shipment->get_pickup_points(00100, '', 'FI', '2103'));
}, 11);
```

image

Adding a catch would not help, as the assert would still fail. There's no problem with the code, the test is just bad. Here's the same test printed to console;

```
add_action('wp_loaded', function() use ($instance) {
  try {
    $x = $instance->shipment->get_pickup_points(00100, '', 'FI', '2103');
  } catch (\Exception $e) {
    $x = $e->getMessage();
  }

  try {
    $z = new Pakettikauppa\Client();
    $y = $z->searchPickupPoints(00100, '', 'FI', '2103');
  } catch (\Exception $e) {
    $y = $e->getMessage();
  }

  error_log(print_r($y, true));
  error_log(print_r($x, true));
}, 11);

// output
[30-Mar-2020 10:07:14 UTC] No pickup points were found. Check the address.
[30-Mar-2020 10:07:15 UTC] Array
(
)

```

Maybe the test should test an address with pickup points instead. `get_pickup_points` doesn't modify the results, so the test **should** work with a non-empty address.

**Comment by k1sul1** [comment]
Merge #147 before merging this.

**Comment by k1sul1** [comment]
I have no idea how to resolve the gnitpick issue, I thought that it would resolve itself by merging #147.

**Comment by ottok** [comment]
Gnitpick complains:

> Fix these fails and try again:
> - d266ea85ede55969d145ec3b9823850c1cca9c04: Commit message title 'Merge branch 'features/api-library-2.0' of https://github.com/Seravo/woo-pakettikauppa into features/api-library-2.0' is over 72 characters
> - d266ea85ede55969d145ec3b9823850c1cca9c04: Merge request includes merges by itself
> - a8e9b2a028540a42048dbf8f7404469089e7b8a1: Commit message title 'Merge branch 'features/api-library-2.0' of https://github.com/Seravo/woo-pakettikauppa into features/api-library-2.0' is over 72 characters
> - a8e9b2a028540a42048dbf8f7404469089e7b8a1: Merge request includes merges by itself
> - 1eb50030df0d55073de8443a01a5523a507e5778: Commit message title 'Merge branch 'master' of https://github.com/Seravo/woo-pakettikauppa into features/api-library-2.0' is over 72 characters
> - 1eb50030df0d55073de8443a01a5523a507e5778: Merge request includes merges by itself
> - 5e4c9870500b0a5c0d20e132fde0f56d946ed551: Merge request includes merges by itself

This "Merge request includes merges by itself" stems from the fact that you are merging on master in the PR itself, when you should rebase on master.

Now the history looks like this which makes the code base harder to maintain since tracking changes (and fixing regressions) will take twice as much time:
image

Also reviewing a PR that changes 28 files in 20 commits is hard.

See https://github.com/Seravo/gnitpick/blob/master/README.md

**Comment by ottok** [comment]
Gnitpick complains:

> Fix these fails and try again:
> - d266ea85ede55969d145ec3b9823850c1cca9c04: Commit message title 'Merge branch 'features/api-library-2.0' of https://github.com/Seravo/woo-pakettikauppa into features/api-library-2.0' is over 72 characters
> - d266ea85ede55969d145ec3b9823850c1cca9c04: Merge request includes merges by itself
> - a8e9b2a028540a42048dbf8f7404469089e7b8a1: Commit message title 'Merge branch 'features/api-library-2.0' of https://github.com/Seravo/woo-pakettikauppa into features/api-library-2.0' is over 72 characters
> - a8e9b2a028540a42048dbf8f7404469089e7b8a1: Merge request includes merges by itself
> - 1eb50030df0d55073de8443a01a5523a507e5778: Commit message title 'Merge branch 'master' of https://github.com/Seravo/woo-pakettikauppa into features/api-library-2.0' is over 72 characters
> - 1eb50030df0d55073de8443a01a5523a507e5778: Merge request includes merges by itself
> - 5e4c9870500b0a5c0d20e132fde0f56d946ed551: Merge request includes merges by itself

This "Merge request includes merges by itself" stems from the fact that you are merging on master in the PR itself, when you should rebase on master.

Now the history looks like this which makes the code base harder to maintain since tracking changes (and fixing regressions) will take twice as much time:
image

Also reviewing a PR that changes 28 files in 20 commits is hard.

See https://github.com/Seravo/gnitpick/blob/master/README.md

**Comment by joosev** [comment]
I don't know where those come from. I use git rebase master if I need to use these days. And I guess those git merges can not be fixed anymore because they are in there already?

**Comment by joosev** [comment]
Are you saying that Pull Requests should not be merged to master but rebased on master?

**Comment by joosev** [comment]
Instead of "create merge commit" we should use "rebase and merge"?

**Comment by ottok** [comment]
> I don't know where those come from.

Look at the git history and everything is transparent. You are still not using rebases. Look at the pink line here:

image

It should branch of master and then when the PR is closed merge back in. Now it starts off from some very old version of master.

Page https://github.com/Seravo/gnitpick/blob/master/README.md links to some very good git docs and presentations.

> Instead of "create merge commit" we should use "rebase and merge"?

You should start by using `gitk --all` or some similar tool that shows you the history. After that it is much easier for you to grasp how git works, what is the state of different repositories and branches etc.

Here is a pretty good step-by-step documentation how to handle branches and how to do a rebase: https://gitlab.com/fevermap/fevermap/-/blob/master/CONTRIBUTING.md#rebasing-a-merge-request

Basically you need to learn the commands `git pull` and `git rebase -i` which you run on your own computer to manage the branch you have yourself.
